internal/traceprof/traceproftest/app.go

```diff
@@ -0,0 +1,276 @@
+		a.grpcServer = grpc.NewServer(grpc.StreamInterceptor(si), grpc.UnaryInterceptor(ui))
+		pb.RegisterTestAppServer(a.grpcServer, a)
+		go a.grpcServer.Serve(l)
+		a.grpcClientConn, err = grpc.Dial(l.Addr().String(), grpc.WithInsecure())
+		require.NoError(t, err)
+	case HTTP:
+		router := httptrace.New()
+		// We use a routing pattern here so our test can validate that potential
+		// Personal Identifiable Information (PII) values, in this case :secret,
+		// isn't beeing collected in the "trace endpoint" label.
+		router.Handle("POST", HTTPWorkEndpoint, a.workHandler)
+		a.httpServer = httptest.NewServer(router)
+	default:
+		panic("unreachable")
+	}
+	a.CPUProfiler.start(t)
+}
+
+// Stop stops the app, tracer and cpu profiler in an idempotent fashion.
+func (a *App) Stop(t testing.TB) {
+	if a.stopped {
+		return
+	}
+	a.prof = a.CPUProfiler.Stop(t)
+	tracer.Stop()
+	switch a.config.AppType {
+	case Direct:
+		// nothing to tear down
+	case GRPC:
+		a.grpcServer.GracefulStop()
+		a.grpcClientConn.Close()
+	case HTTP:
+		a.httpServer.Close()
+	default:
+		panic("unreachable")
+	}
+	a.stopped = true
+}
+
+// WorkRequest sends the given req to the demo app and returns the response.
+// The config.AppType determines how the request is made.
+func (a *App) WorkRequest(t testing.TB, req *pb.WorkReq) *pb.WorkRes {
+	switch a.config.AppType {
+	case Direct:
+		res, err := a.Work(context.Background(), req)
+		require.NoError(t, err)
+		return res
+	case GRPC:
+		client := pb.NewTestAppClient(a.grpcClientConn)
+		res, err := client.Work(context.Background(), req)
+		require.NoError(t, err)
+		return res
+	case HTTP:
+		body, err := json.Marshal(req)
+		require.NoError(t, err)
+		url := a.httpServer.URL + "/work/secret-pii"
+		res, err := http.Post(url, "text/plain", bytes.NewReader(body))
+		require.NoError(t, err)
+
+		defer res.Body.Close()
+		workRes := &pb.WorkRes{}
+		require.NoError(t, json.NewDecoder(res.Body).Decode(&workRes))
+		return workRes
+	default:
+		panic("unreachable")
+	}
+}
+
+// CPUProfile stops the app and returns its CPU profile.
+func (a *App) CPUProfile(t testing.TB) *CPUProfile {
+	a.Stop(t)
+	return a.prof
+}
+
+func (a *App) workHandler(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	req := &pb.WorkReq{}
+	if err := json.NewDecoder(r.Body).Decode(req); err != nil {
+		http.Error(w, "bad body", http.StatusBadRequest)
+	}
+	res, err := a.Work(r.Context(), req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	json.NewEncoder(w).Encode(res)
+}
+
+// Work implements the request handler for the demo app. It's reused regardless
+// of the config.AppType.
+func (a *App) Work(ctx context.Context, req *pb.WorkReq) (*pb.WorkRes, error) {
+	// Simulate user-defined custom labels to make sure we don't overwrite them
+	// when we apply our own labels.
+	ctx = pprof.WithLabels(ctx, toLabelSet(CustomLabels))
+	pprof.SetGoroutineLabels(ctx)
```

Comment:
> So if someone else, like a third-party package, sets is own pprof labels too but without using the "current context", it would overwrite the existing ones, right?

Yes. I'm planning to document this in our official docs. Generally speaking I'd say that usage of custom goroutine labels is still rather rare, and the people who use them are aware of the need to propagate contexts if they want things to work smoothly.

> If yes, that would be another example of good use-case for a goroutine local storage to avoid having to pass this Go context to every called function...

Labels are goroutine local storage! It's just write-only storage 🙈 

But yeah, I hear ya! I just don't think it's in the cards as far as the Go project is concerned. I've made my peace with it.